### DISCUSSION TITLE
Benchmarks (hev-socks5-tunnel vs xjasonlyu-tun2socks)

### DISCUSSION THREAD

**Discussion by heiher** [xjasonlyu/tun2socks#216] [General]
It seems that the performance of tun2socks is really good due to the support of multi-threading, but it needs more CPU usage.

## Hardware/Software

CPU: AMD Ryzen 9 7950X 16-Core Processor (Max 4.5GHz)

OS: Arch Linux

### Socks5 server

Repo: https://github.com/heiher/hev-socks5-server

Version: 2.4.7

### Topology

```
Namespace root                      Namespace container
eth0:  192.168.0.8
veth0: 192.168.1.1                  veth1: 192.168.1.2
socks5 server(listen: 192.168.1.1)  hev-socks5-tunnel(connect: 192.168.1.1)
                                    tun2socks(connect: 192.168.1.1)
iperf3 server(listen: 192.168.0.8)  iperf3 client(connect: 192.168.0.8)
```

### MTU

eth0/veth0/veth1: 1500
tun0: 9000

## hev-socks5-tunnel

Repo: https://github.com/heiher/hev-socks5-tunnel

Version: 2.4.3

#### Upload

```
$ iperf3 -c 192.168.0.8  
- - - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - -
[  5]   0.00-10.00  sec  23.0 GBytes  19.8 Gbits/sec    0             sender
[  5]   0.00-10.04  sec  23.0 GBytes  19.7 Gbits/sec                  receiver
CPU usage: 76%
MEM usage: 2.7M
```

```
$ iperf3 -c 192.168.0.8 -P 10
- - - - - - - - - - - - - - - - - - - - - - - - -
[SUM]   0.00-10.00  sec  21.1 GBytes  18.1 Gbits/sec    0             sender
[SUM]   0.00-10.50  sec  21.1 GBytes  17.3 Gbits/sec                  receiver
CPU usage: 90%
MEM usage: 3.3M
```

#### Download

```
$ iperf3 -c 192.168.0.8 -R 
- - - - - - - - - - - - - - - - - - - - - - - - -
[  5]   0.00-10.04  sec  24.8 GBytes  21.2 Gbits/sec    0             sender       
[  5]   0.00-10.00  sec  24.7 GBytes  21.3 Gbits/sec                  receiver  
CPU usage: 90%                                                                  
MEM usage: 2.7M 
```

```
$ iperf3 -c 192.168.0.8 -R -P 10
- - - - - - - - - - - - - - - - - - - - - - - - -
[SUM]   0.00-10.04  sec  24.7 GBytes  21.1 Gbits/sec    0             sender       
[SUM]   0.00-10.00  sec  24.5 GBytes  21.1 Gbits/sec                  receiver  
CPU usage: 90%                                                                  
MEM usage: 3.2M
```

## tun2socks

Repo: https://github.com/xjasonlyu/tun2socks

Version: 2.4.1

### Upload

```
$ iperf3 -c 192.168.0.8  
- - - - - - - - - - - - - - - - - - - - - - - - -
[  5]   0.00-10.00  sec  6.15 GBytes  5.28 Gbits/sec    0             sender
[  5]   0.00-10.04  sec  6.15 GBytes  5.26 Gbits/sec                  receiver
CPU usage: 280%
MEM usage: 19M
```

```
$ iperf3 -c 192.168.0.8 -P 10
- - - - - - - - - - - - - - - - - - - - - - - - -
[SUM]   0.00-10.00  sec  5.74 GBytes  4.93 Gbits/sec  5386             sender
[SUM]   0.00-10.04  sec  5.72 GBytes  4.89 Gbits/sec                  receiver
CPU usage: 320%
MEM usage: 23M
```

#### Download

```
$ iperf3 -c 192.168.0.8 -R 
- - - - - - - - - - - - - - - - - - - - - - - - -
[  5]   0.00-10.04  sec  5.37 GBytes  4.60 Gbits/sec    0             sender
[  5]   0.00-10.00  sec  5.36 GBytes  4.61 Gbits/sec                  receiver
CPU usage: 220%
MEM usage: 20M
```

```
$ iperf3 -c 192.168.0.8 -R -P 10
- - - - - - - - - - - - - - - - - - - - - - - - -
[SUM]   0.00-10.24  sec  22.9 GBytes  19.2 Gbits/sec  272             sender
[SUM]   0.00-10.00  sec  22.2 GBytes  19.1 Gbits/sec                  receiver
CPU usage: 600%
MEM usage: 74M
```

**Comment by xjasonlyu** [comment]
It's inevitable to consume more CPU since tun2socks uses a user-space netstack and multithreading underlying. BTW, gVisor is constantly updating its tcpip stack and improving performance, so I would not worry about the high CPU usage for now.

**Reply by heiher** [reply]
Try to running your tun2socks on single CPU thread, It looks faster than on multi-threads for some cases.

**Comment by heiher** [comment]
# CPU usage on full 1Gbps

## Hardware/Software

### Server

CPU: AMD Ryzen 9 7950X 16-Core Processor (Max 4.5GHz)

OS: Arch Linux

#### Socks5 server

Repo: https://github.com/heiher/hev-socks5-server

Version: 2.4.7

### Client

CPU: Intel(R) Core(TM) i7-7700K CPU @ 4.20GHz

OS: Arch Linux

#### tun2socks

Version: 2.4.1

#### hev-socks5-tunnel

Version: 2.2.0

### Link bandwidth

Server <-> client: 1Gbps

## CPU usage

![cpu-usage-1g](https://user-images.githubusercontent.com/1407733/213208501-e52d9181-24ed-49d4-9936-8903b4204662.png)


### tun2socks

```
Upload   signle: 100%
Upload   multi : 100%
Download single: 110%
Download multi:  110%
```

### hev-socks5-tunnel

```
Upload   signle: 40%
Upload   multi : 40%
Download single: 25%
Download multi:  30%
```

**Comment by heiher** [comment]
[Here](https://github.com/heiher/hev-socks5-tunnel/wiki/Benchmarks) are benchmarks (speed and cpu usage) of tun2socks/go-tun2socks/badvpn-tun2socks and hev-socks5-tunnel on MTU 8500.

### Speed

![](https://github.com/heiher/hev-socks5-tunnel/wiki/res/upload-speed.png)
![](https://github.com/heiher/hev-socks5-tunnel/wiki/res/download-speed.png)

### CPU usage

![](https://github.com/heiher/hev-socks5-tunnel/wiki/res/upload-cpu.png)
![](https://github.com/heiher/hev-socks5-tunnel/wiki/res/download-cpu.png)

### Memory usage

![](https://github.com/heiher/hev-socks5-tunnel/wiki/res/upload-mem.png)
![](https://github.com/heiher/hev-socks5-tunnel/wiki/res/download-mem.png)

**Comment by heiher** [comment]
## xjasonlyu-tun2socks

Repo: https://github.com/xjasonlyu/tun2socks

Version: 2.6.0

Command:

```bash
tun2socks-linux-amd64 -device tun0 -mtu 8500 -proxy socks5://192.168.1.1:1080
```

### Upload

```
$ iperf3 -c 192.168.0.8
- - - - - - - - - - - - - - - - - - - - - - - - -
[  5]   0.00-10.00  sec  17.8 GBytes  15.3 Gbits/sec    0            sender
[  5]   0.00-10.04  sec  17.8 GBytes  15.2 Gbits/sec                 receiver
CPU usage: 163%
MEM usage: 22.1M
```

```
$ iperf3 -c 192.168.0.8 -P 10
- - - - - - - - - - - - - - - - - - - - - - - - -
[SUM]   0.00-10.00  sec  30.0 GBytes  25.8 Gbits/sec  437             sender
[SUM]   0.00-10.04  sec  30.0 GBytes  25.7 Gbits/sec                  receiver
CPU usage: 296%
MEM usage: 32.5M
```

### Download

```
$ iperf3 -c 192.168.0.8 -R
- - - - - - - - - - - - - - - - - - - - - - - - -
[  5]   0.00-10.04  sec  13.6 GBytes  11.7 Gbits/sec    1            sender
[  5]   0.00-10.00  sec  13.6 GBytes  11.7 Gbits/sec                 receiver
CPU usage: 144%
MEM usage: 25.1M
```

```
$ iperf3 -c 192.168.0.8 -R -P 10
- - - - - - - - - - - - - - - - - - - - - - - - -
[SUM]   0.00-10.04  sec  33.2 GBytes  28.4 Gbits/sec  256             sender
[SUM]   0.00-10.00  sec  33.1 GBytes  28.4 Gbits/sec                  receiver
CPU usage: 511%
MEM usage: 49.1M
```

A marked improvement! :+1:

**Comment by Plangto** [comment]
i actually wanted to use hev socks5 tunnel but its setup is so hard while tun2socks is very easy to setup
